game/src/ecs/components/ai/AITools.java

```diff
@@ -0,0 +1,121 @@
+package ecs.components.ai;
+
+import com.badlogic.gdx.ai.pfa.GraphPath;
+import ecs.components.PositionComponent;
+import ecs.components.VelocityComponent;
+import ecs.entities.Entity;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import level.elements.ILevel;
+import level.elements.tile.Tile;
+import level.tools.Coordinate;
+import mydungeon.ECS;
+import tools.Point;
+
+public class AITools {
+    private static Random random = new Random();
+
+    public static GraphPath<Tile> calculateNewPath(Entity entity, float radius) {
+        PositionComponent pc = (PositionComponent) entity.getComponent(PositionComponent.name);
+        VelocityComponent vc = (VelocityComponent) entity.getComponent(VelocityComponent.name);
+        if (pc != null && vc != null) {
+            ILevel level = ECS.currentLevel;
+            Point position = pc.getPosition();
+            Tile currentTile = level.getTileAt(position.toCoordinate());
+            List<Tile> tiles = new ArrayList<>();
+            for (float x = position.x - radius; x <= position.x + radius; x++) {
+                for (float y = position.y - radius; y <= position.y + radius; y++) {
+                    tiles.add(level.getTileAt(new Point(x, y).toCoordinate()));
+                }
+            }
+            tiles.removeIf(tile -> tile == null);
+            tiles.removeIf(tile -> !tile.isAccessible());
+            Coordinate newPosition = tiles.get(random.nextInt(tiles.size())).getCoordinate();
+            GraphPath path =
+                    level.findPath(
+                            level.getTileAt(position.toCoordinate()), level.getTileAt(newPosition));
+
+            System.out.println("PATH " + path.getCount());
+            return path;
+        }
+        return null;
```

Comment:
Was bedeutet `null` als Ergebnis? Die Operation hat nicht geklappt, weil es keine `PositionComponent` und/oder `VelocityComponent` gab? Wäre das nicht die Stelle, eine passende Exception auszulösen (dito in `PositionComponent` und `VelocityComponent`)? Dann braucht der Aufrufer nicht die ganzen `if (wuppie != null) { ...} else {return null;}` ...

Reply:
Jop. Das pattern gibt es häufiger und die Problematik lässt sich auf die Natur eines ECS zurückführen. Ich wollte nochmal in der großen Welt gucken, wie andere damit umgehen. 

Dauer Lösung ist es so nicht. 
Ist ja auch nur draft

Reply:
Schon klar. Leider hält ein Provisorium oft länger, als einem lieb ist ;)

Ich denke, die Lösung heißt `Exception` oder `Option<T>`, wobei letzteres einen etwas anderen Code-Stil nach sich zieht (oder Du endest dann analog zu den `if (null)`-Abfragen mit `.isEmpty()`).

Reply:
Schau mal https://github.com/Programmiermethoden/PM-Lecture/blob/master/markdown/modern-java/optional.md#lsf-liefert-jetzt-optional-zurück und https://github.com/Programmiermethoden/PM-Lecture/blob/master/markdown/modern-java/optional.md#zugriff-auf-optional-objekte: das passt eigentlich exakt zu dem Problem hier, der Zugriff dann mit `best = lsf.getBestStudi().orElseThrow();`.

Reply:
Das wäre eine Möglichkeit, welches sich am aktuellen Konzept orientiert.

Ich weiß nur nicht, ob ein null-Check (in welcher Form auch immer) überhaupt notwendig ist. 
Vielleicht sollte das Programm hier crashen.

Immerhin hab ich (als Anwender) in meiner Entität eine  Funktion hinterlegt (über das `AIComponent`), welches nunmal das `PositionComponent` und `VelocityComponent` Voraussetzt.

Aber so wirklich richtig klingt das auch nicht.

Reply:
- Ich bin immer ein Fan davon, möglichst robusten Code zu schreiben und vorhersehbare Probleme zu berücksichtigen.
- Fehler des Anwenders kann ich nur bedingt auffangen - ein Crash scheint in Deinem Szenario unvermeidlich.
- Ein Crash darf aber informativ sein und den Nutzer in die Lage versetzen zu erkennen, was er falsch gemacht hat bzw. was von ihm erwartet wird - wenn es hilfreiche Fehlermeldungen von uns gibt.
- Ein Teil des Problems sind aber wir selbst, wenn wir "null" als Initialisierung und/oder Rückgabewerte überhaupt erlauben. Das ist zwar bequem, führt aber den ganzen Problemen bzw. unterstützt diese. Und es gibt ja moderne   Mittel und Wege, vernünftiger damit umzugehen.

Reply:
Ich glaube, Du willst wirklich durchgängig mit Option arbeiten. 

Und wenn etwas nicht da ist (aber da sein sollte), dann ist das imho ein Fehlerzustand. Und die Antwort in Java auf Fehler ist nicht null, sondern eine Exception.

Reply:
> Und wenn etwas nicht da ist (aber da sein sollte), dann ist das imho ein Fehlerzustand. Und die Antwort in Java auf Fehler ist nicht null, sondern eine Exception.

Oder es ist ein Fehler im System-Design. Das ist die andere Möglichkeit. 

Wenn wir null weiter "erlauben", dann ist Optional the way to go.

Reply:
ja, das zieht sich ja durch das ganze design und muss gut überlegt werden